### ISSUE TITLE
inheritance when dataclass is used

### ISSUE THREAD

**Issue by cloasdata** [sqlalchemy/sqlalchemy#9563]
### Describe the bug

Seems like sa falls over the well known dataclass bug with default values in parent class as explained here: https://stackoverflow.com/questions/69711886/python-dataclasses-inheritance-and-default-values

I did not find any 

### Optional link from https://docs.sqlalchemy.org which documents the behavior that is expected

https://docs.sqlalchemy.org/en/20/orm/inheritance.html

### SQLAlchemy Version in Use

2.0.2

### DBAPI (i.e. the database driver)

pysqllite

### Database Vendor and Major Version

SQLite

### Python Version

3.11.2

### Operating system

Windows

### To Reproduce

```python
from sqlalchemy import ForeignKey
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import MappedAsDataclass


class Base(MappedAsDataclass, DeclarativeBase):
    pass


class Employee(Base):
    __tablename__ = "employee"
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_columns(default="42")
    type: Mapped[str]

    __mapper_args__ = {
        "polymorphic_identity": "employee",
        "polymorphic_on": "type",
    }

    def __repr__(self):
        return f"{self.__class__.__name__}({self.name!r})"

class Engineer(Employee):
    __tablename__ = "engineer"
    id: Mapped[int] = mapped_column(ForeignKey("employee.id"), primary_key=True)
    engineer_name: Mapped[str]

    __mapper_args__ = {
        "polymorphic_identity": "engineer",
    }
```

### Error

```
# Copy the complete stack trace and error message here, including SQL log output if applicable.
```

**Comment by zzzeek** [comment]
is this the error?

    TypeError: non-default argument 'type' follows default argument

that's expected

you would need to remove these items from the init list, use kw_only as needed, or disable the init method.

example:

```py
from sqlalchemy import ForeignKey
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import MappedAsDataclass


class Base(MappedAsDataclass, DeclarativeBase):
    pass


class Employee(Base):
    __tablename__ = "employee"
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(default="42")
    type: Mapped[str] = mapped_column(init=False)

    __mapper_args__ = {
        "polymorphic_identity": "employee",
        "polymorphic_on": "type",
    }

    def __repr__(self):
        return f"{self.__class__.__name__}({self.name!r})"


class Engineer(Employee):
    __tablename__ = "engineer"
    id: Mapped[int] = mapped_column(
        ForeignKey("employee.id"), primary_key=True
    )
    engineer_name: Mapped[str] = mapped_column(kw_only=True)

    __mapper_args__ = {
        "polymorphic_identity": "engineer",
    }

```

see https://docs.python.org/3/library/dataclasses.html#inheritance for how dataclasses work with inheritance

**Comment by cloasdata** [comment]
Okay it is expected. But there is no hint or something in the docs when you fall over and not having the whole dataclass doc in mind ;-)
Which leads to: Oh my code must be faulty. Until you absolute sure and you decide to ask google about inheritance with dataclasses. And that could take time :)

Thanks anyway for the explanation!

**Comment by CaselIT** [comment]
I'm not sure what you are suggesting SQLAlchemy do here? Catch the type error and re-raise it saying something like:
"Cannot construct dataclass because of xyz. Check the dataclass documentation at https://docs.python.org/3/library/dataclasses.html" ?

I think this we can do

**Comment by cloasdata** [comment]
That would be more than I could think of. 
It would be already nice if you could find it in the documentation.

**Comment by CaselIT** [comment]
@zzzeek what's your opinion on wrapping dataclass errors to re-raise them adding some additional information?

**Comment by zzzeek** [comment]
Im sort of -1 on that.   We can add dragon notes to the docs, "please refer to dataclass docs for typeerrors" or "notes on inheritance, please see dataclasses at .."

**Comment by zzzeek** [comment]
not really sure really.   that this is "real dataclasses" is what's supposed to make this straightforward

**Comment by CaselIT** [comment]
yes, but it's not the first time this or similar dataclass limitations gets reported, so maybe people are not that familiar with them?

In any case i've mentioned this issue in #9410 regarding re-raise we can decide later, since it's easy to add, not that easy to remove

**Comment by zzzeek** [comment]
yeah of course.   guess it will help

**Comment by CaselIT** [comment]
> why are people using the dataclass thing without knowing how to use dataclasses?

don't know. maybe new & shiny ? I don't think the sqlalchemy docs mention anything like "dataclass is the way" or "dataclass will be the way"

**Comment by sqla-tester** [comment]
**Federico Caselli** has proposed a fix for this issue in the **main** branch:

**Wrap dataclass exceptions clarifying origin** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4535
